### ISSUE TITLE
Optimized

### ISSUE THREAD

**Issue by jasonjin34** [MouYongli/DeepSC#16]
1. GPU optimization for ensuring that no gpu vram waste, max utils
2. dataloader optimization for this part of code https://github.com/MouYongli/DeepSC/blob/65c7b3d3d1405dc3336bfdcf2b613e8ef38760ce/src/deepsc/train/trainer.py#L72-L82
check the following link
https://docs.pytorch.org/docs/stable/data.html
3. logging with wandb use only one GPU
4. increase saving checkpoint frequency (every one hour or 30 mins)
5. need to implement the resume training function
https://lightning.ai/docs/pytorch/stable/common/checkpointing_basic.html

**Comment by jasonjin34** [comment]
6. optimization gradient accumulating https://lightning.ai/docs/fabric/2.4.0/advanced/gradient_accumulation.html
7.  save checkpoint https://github.com/MouYongli/DeepSC/blob/c036f36ff07a773291dc75b1b86967c2ade1f18e/src/deepsc/utils/utils.py#L133-L148
replace with this 
https://lightning.ai/docs/fabric/2.4.0/guide/checkpoint/distributed_checkpoint.html

**Comment by LeonWYXFEV** [comment]
1. GPU utility and memory maxized
2. 效果很大，提升到了47m/epoch (together with model compile)
3. ✅
4. ✅
5. ✅
6: 已实现，但提升有限
7. ✅
8. model.compile not compatible to scBERT. Will crash at the end of one epoch
